enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,264 @@
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name? 
+- In which namespace the new operator components should be hosted?
+
+### current implmentation 
+In current implementation, the [networking infra componenets](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md) run as static pods through the MCO.
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following [diagram](https://docs.google.com/drawings/d/1a8K16vcaBB6SIF0bN8DaTj4ltjBxVsVNjJN5NCd-U0A/edit?usp=sharing) reflects the suggested changes.
+
+#### Self hosted DNS services
+- The self hosted DNS functionality should be composed of the following components in each node:
+    - A NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+    - Systemd service running dnsmasq
+    - CoreDNS-mDNS pod
+- The CHNSO will be responsible only for the deployment of the CoreDNS pods while the NetworkManager dispatcher script and dnsmasq Systemd service will run through MCO.
+- To avoid port conflict, Dnsmasq service should listen on DNS well-known port (53) and CoreDNS-MDNS should listen on other port (e.g: 5353). 
+- The new Dnsmasq service should be responsible to resolve 'api-int' (to API-VIP value provided in install-config) hostname when CoreDNS-MDNS pod isn't running on the node.
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO as DaemonSet pods instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+
+#### Master node shutdown
+In order not to break the [graceful termination process of kube-api](https://github.com/openshift/installer/blob/master/docs/dev/kube-apiserver-health-check.md) the HAProxy pod should be stopped at the same time or after the kube-api static pod.
+
+Using [K8S graceful node shutdown in kubelet](https://github.com/kubernetes/kubernetes/pull/96129) it should be possible to define that HAProxy pod should be terminated after kube-api, by setting HAProxy pod a higher priority than the kube-api static pod.
```

Comment:
This is still so much hand-waving :-/

The kube-apiserver pod has `priorityClassName: system-node-critical`. With the linked enhancement there is no "more critical than critical". Hence, we cannot order haproxy to be shutdown after kube-apiserver.

tl/dr: this is still a blocker.

Be precise here. Remove the hand-waving and be super clear how each pod is configured.

Reply:
A. Following the discussion on slack with @smarterclayton, I understood that it should be possible to delete the HAProxy pod after (or at the same time as ) the kube-apiserver.

B. Is this ordering problem relevant only for HAProxy and kube-apiserver ? 
I assume that there are other non-static pods (sdn pod ?) that need to be deleted after static pods.   Can't we use the solution that will be used for other similar cases also for the HAProxy - API? 

C. Given that the K8S graceful node shutdown in kubelet feature under Openshift is still under discussion/development, I do not think it is possible to provide a detailed configuration of pods at this point

Reply:
I think the important ordering here is to ensure that keepalived shuts down before haproxy so we don't break connections twice. Once keepalived stops and the VIP moves to another node it doesn't matter what happens with haproxy or the apiserver on the node that is shutting down. Neither one should be receiving connections after that.

Reply:
> I think the important ordering here is to ensure that keepalived shuts down before haproxy so we don't break connections twice. Once keepalived stops and the VIP moves to another node it doesn't matter what happens with haproxy or the apiserver on the node that is shutting down. Neither one should be receiving connections after that.

Yep, that's correct. the pods should be deleted in the following order kube-api,Keepalived and HAProxy.

Question: Isn't it possible to force deletion of kube-api, Keepalived , and HAProxy in the desired order simply by implementing smart delay code in SIGTERM handler in Keepalived and HAProxy pods?

Hmmm, I'm wondering what's the added value for implementing the api graceful termination for shutdown of node holds the VIP (master-x)  in case of self-hosted LB.
While we allow graceful termination of existing connections to master-x, all new connections to nodes !=x that will be created during graceful shutdown period will break upon VIP switchover.
 
I can understand the added value with graceful termination for the following cases  :
 -  external LB 
 -  Shutdown of a node not holding the VIP - self-hosted LB

Am I missing something here?

Reply:
> Yep, that's correct. the pods should be deleted in the following order kube-api,Keepalived and HAProxy.

Not necessarily. We don't care when kube-api actually goes away, only when it flips the readyz endpoint to red. As soon as that happens we'll stop sending traffic to it. At that point connections will still be coming in to keepalived, sent to haproxy, and then to the other two masters.

However, see below because that part is what I was saying in my first comment.

> 
> Question: Isn't it possible to force deletion of kube-api, Keepalived , and HAProxy in the desired order simply by implementing smart delay code in SIGTERM handler in Keepalived and HAProxy pods?
> 
> Hmmm, I'm wondering what's the added value for implementing the api graceful termination for shutdown of node holds the VIP (master-x) in case of self-hosted LB.
> While we allow graceful termination of existing connections to master-x, all new connections to nodes !=x that will be created during graceful shutdown period will break upon VIP switchover.

Right, this is what I was getting at. We can't avoid the connection break on VIP move. We _can_ minimize the damage by making sure we don't also break on haproxy shutdown.

If/when we have an internal loadbalancer that can gracefully fail over on node reboot we'll need to revisit this, but there will almost certainly be different services involved then so it's impossible to design for that today.

So I think the main point here is to ensure that keepalived stops before haproxy. What kube-apiserver is doing in the meantime is actually irrelevant because our services don't support the graceful shutdown needed anyway.